src/QsCompiler/QirGeneration/DebugInfoManager.cs

```diff
@@ -0,0 +1,575 @@
+
+        /// <summary>
+        /// Contains the location information for the namespace element we are inside
+        /// </summary>
+        internal QsLocation? CurrentNamespaceElementLocation { get; set; }
+
+        /// <summary>
+        /// The GenerationContext that owns this DebugInfoManager
+        /// </summary>
+        private readonly GenerationContext sharedState;
+
+        /// <summary>
+        /// Contains the DIBuilders included in the module related to this DebugInfoManager.
+        /// The key is the source file path of the DebugInfoBuilder's compile unit
+        /// </summary>
+        private readonly Dictionary<string, DebugInfoBuilder> dIBuilders;
+
+        /// <summary>
+        /// Constructs a DebugInfoManater with access to the <see cref="GenerationContext"/> as the shared state.
+        /// </summary>
+        internal DebugInfoManager(GenerationContext generationContext, bool debugSymbolsEnabled)
+        {
+            this.sharedState = generationContext;
+            Config.DebugSymbolsEnabled = debugSymbolsEnabled;
+            this.StatementLocationStack = new Stack<QsLocation?>();
+            this.ExpressionRangeStack = new Stack<DataTypes.Range?>();
+            this.dIBuilders = new Dictionary<string, DebugInfoBuilder>();
+        }
+
+        /// <summary>
+        /// If <see cref="Config.DebugSymbolsEnabled"/> is set to true, creates and adds top level debug info to the module.
+        /// Note: because this is called from within the constructor of the GenerationContext,
+        /// we cannot access this.Module or anything else that uses <see cref="sharedState"/>.
+        /// </summary>
+        /// <param name="owningModule">The <see cref="BitcodeModule"/> that this DebugInfoManager is related to</param>
+        internal void AddTopLevelDebugInfo(BitcodeModule owningModule)
+        {
+            if (!Config.DebugSymbolsEnabled)
+            {
+                return;
+            }
+
+            // Add Module identity and Module Flags
+            owningModule.AddProducerIdentMetadata(Config.GetQirProducerIdent());
+            owningModule.AddModuleFlag(ModuleFlagBehavior.Warning, BitcodeModule.DwarfVersionValue, Config.DwarfVersion);
+            owningModule.AddModuleFlag(ModuleFlagBehavior.Warning, BitcodeModule.DebugVersionValue, BitcodeModule.DebugMetadataVersion);
+            owningModule.AddModuleFlag(ModuleFlagBehavior.Warning, Config.CodeviewName, Config.CodeviewVersion);
+        }
+
+        /// <summary>
+        /// Makes the necessary calls to finalize the <see cref="DebugInfoBuilder"/>s.
+        /// </summary>
+        internal void FinalizeDebugInfo()
+        {
+            if (!Config.DebugSymbolsEnabled)
+            {
+                return;
+            }
+
+            foreach (KeyValuePair<string, DebugInfoBuilder> entry in this.dIBuilders)
+            {
+                entry.Value.Finish(); // must be called for every DIBuilder after all QIR generation
```

Comment:
If this clean up has to happen when DebugInfoManager is done being used, should it be part of an IDisposable implementation for DebugInfoManager instead?

Reply:
I tried this, but the problem is that the only place it would make sense (to me) to call Dispose would be within the GenerationContext when it is being disposed. (Since the GenerationContext is paired 1-1 with a DebugInfoManager and the GenerationContext is the shared state for the transformation. However, FinalizeDebugInfo needs to be called before we emit the QIR which we do before we dispose of the GenerationContext. 

Calling Finish() on each DIBuilder is so that any unresolved dependencies in the debug info tags are resolved instead of being left as "temporary" (this is all related to the LLVM Bindings). This feels less about managed resources, and more about signaling to the LLVM Bindings that we are finished parsing the functions in the source file and ready to do the final wrap up.

Let me know if you think it should be handled differently.

Reply:
Thanks for the explanation. That sounds fine, it can stay the way it is.